### PR TITLE
Reusable Dropdowns

### LINKED ISSUE #3494: Difficult to tell which source is currently selected in source selector based on styles
screen shot 2018-05-21 at 11 57 10 am

### PR DESCRIPTION
**Pull request by alexpaxton**
Closes #3494 

This PR is a subset of the work listed in https://github.com/influxdata/chronograf/issues/3946

### Composable API

Define your dropdown menu as an array of objects
```
interface MenuItem {
  label: string
  id: string
  hex: string
}

menuItems: MenuItem[] = [
  {label: 'red', id: 'color-red', hex: '#ff0000'},
  {label: 'green', id: 'color-green', hex: '#00ff00'},
  {label: 'blue', id: 'color-blue', hex: '#0000ff'},
]
```
Then map over the array to pass in the children:
```
<Dropdown
  selectedID={menuItems[0].id}
  onChange={this.handleDropdownChange}>
  {menuItems.map(item => (
    <Dropdown.Item key={item.id} id={item.id} value={item}>
      {item.label}
    </Dropdown.Item>
  )}
</Dropdown>
```

`Dropdown.Item` returns the `value` prop when clicked, so I recommend passing the entire object into that prop. The `id` prop is used find and highlight the active item in the list.

```
private handleDropdownChange = (item: MenuItem): void => {
  // Do something with the returned item
  this.setState({selectedItem: item.id})
}
```

### Customization
Here's a breakdown of the optional props. `Dropdown` takes enums which will be shared across the entire set of reusable components. These are `ComponentSize`, `ComponentColor`, and `IconFont`

```
widthPixels:   number              (Controls the pixel width of the dropdown, defaults to 120)
icon:          IconFont            (Displays an icon on the dropdown button, defaults to null)
wrapText:      boolean             (If true the width of the menu with be the same as the button and its contents will wrap.If false the menu width is determined by its contents.Defaults to false)
maxMenuHeight: number              (The height at which the menu will start scrolling, defaults to 250)
customClass:   string              (In case you need to customize positioning or the appearance of dropdowns, defaults to null)
disabled:      boolean             (if true { user } interaction is blocked)
menuColor:     DropdownMenuColors  (defaults to Sapphire.There are 4 options: Sapphire, Malachite, Amethyst, and Onyx)
buttonColor:   ComponentColor      (defaults to Default)
buttonSize:    ComponentSize       (defaults to Small)
```

If you want to add in dividers to the menu I recommend adding in items in your array in which the `text` (in this example) is set to `divider` and then when mapping over the array set a conditional to render a `<Dropdown.Divider />` component instead.

`<Dropdown.Item>` takes children and can be very versatile. This will come in handy when creating dropdowns such as the color selector dropdowns in the CEO, or potentially other use cases.

dropdown-example
In this example I've included a color swatch element as a child of the dropdown items.

### New Patterns
I would like to use these more in general, but especially in the `reusable_ui` set of components

1. Access primary and subcomponents with a single import statement
2. Composable API
3. Enforcing child types

### Checklist
  - [x] Rebased/mergeable
  - [x] Tests pass

### CODE REVIEW

**Comment by alexpaxton** [comment]
TODO: Finish writing test for component

**Comment by alexpaxton** [comment]
TODO:

- [ ] Rename `size` to `buttonSize` as it only affects the button
- [ ] Rename `width` to `widthPixels` to make the values passed in less magical

File: ui/src/reusable_ui/components/dropdowns/Dropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        <div
+          className={`dropdown--menu-container dropdown--${menuColor}`}
+          style={this.menuStyle}
+        >
+          <FancyScrollbar
+            autoHide={false}
+            autoHeight={true}
+            maxHeight={maxMenuHeight}
+          >
+            <div className="dropdown--menu">
```
Comment (chnn) [inline ui/src/reusable_ui/components/dropdowns/Dropdown.tsx]: Could we do the cloning by invoking a new `DropdownItem` component for each child? That would be more explicit and also typecheckable. Something like: 

```jsx
            <div className="dropdown--menu">
              {React.Children.map(children, child => {
                if (!(child instanceof DropdownItem)) {
                  throw new Error('Expected `Dropdown.Item`')
                }

                return (
                  <DropdownItem
                    {...child.props}
                    key={child.props.itemKey}
                    selected={child.props.itemKey === selectedItemKey}
                    onClick={this.handleItemClick}
                  />
                )
              })}
            </div>
```

The thrown error is a bonus to help guide consumers of the `Dropdown` during development.

File: ui/src/reusable_ui/components/dropdowns/Dropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+// Styles
+import './Dropdown.scss'
+
+import {ErrorHandling} from 'src/shared/decorators/errors'
+
+interface Props {
+  children: JSX.Element[]
+  onChange: (value: any) => void
+  selectedItemKey: string
```
Comment (chnn) [inline ui/src/reusable_ui/components/dropdowns/Dropdown.tsx]: See later comment about `key`s and `itemKey`s... think this can just be `selectedKey`.

File: ui/src/reusable_ui/components/dropdowns/Dropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    onChange(value)
+    this.collapseMenu()
+  }
+
+  private validateChildren = (): void => {
+    const {children} = this.props
+
+    if (React.Children.count(children) === 0) {
+      throw new Error(
+        'Dropdowns require at least 1 child element. We recommend using Dropdown.Item and/or Dropdown.Divider.'
```
Comment (chnn) [inline ui/src/reusable_ui/components/dropdowns/Dropdown.tsx]: This is great, seems super helpful for people using the component

File: ui/src/reusable_ui/components/dropdowns/DropdownItem.tsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import React, {Component} from 'react'
+import classnames from 'classnames'
+
+// Types
+import {DropdownChild} from 'src/reusable_ui/types'
+
+import {ErrorHandling} from 'src/shared/decorators/errors'
+
+interface Props {
+  itemKey: string
```
Comment (chnn) [inline ui/src/reusable_ui/components/dropdowns/DropdownItem.tsx]: I think you might be able to just use the `key` prop that's also being passed in. It's confusing to me to have both—is it necessary?

File: ui/src/reusable_ui/types/index.ts [hunk 5]
```diff
@@ -7,6 +7,19 @@ export enum ComponentColor {
...
+  Malachite = 'malachite',
+  Sapphire = 'sapphire',
+  Onyx = 'onyx',
+}
+
+export type DropdownChild =
+  | Array<string | JSX.Element | Element>
+  | string
+  | JSX.Element
+  | Element
```
Comment (chnn) [inline ui/src/reusable_ui/types/index.ts]: lol nice

File: ui/src/reusable_ui/types/index.ts [hunk 6]
```diff
@@ -43,3 +56,72 @@ export enum Greys {
   Ghost = '#fafafc',
   White = '#ffffff',
 }
+
+export enum IconFont {
```
Comment (chnn) [inline ui/src/reusable_ui/types/index.ts]: this is going to be so helpful

**Comment by alexpaxton** [comment]
@chnn I experimented a bit and I found that I can check child types by doing `child.type === DropdownItem` assuming `child` is of type `JSX.Element`

File: ui/src/reusable_ui/components/dropdowns/Dropdown.tsx (hunk 1 above)
Comment (alexpaxton) [reply ui/src/reusable_ui/components/dropdowns/Dropdown.tsx]: This is a great idea -- really makes it clear to the dev how to implement the component properly

**Comment by alexpaxton** [comment]
I replaced `itemKey` with `key` and it works fine when used, however my test fails and gives me an error with this link: https://reactjs.org/warnings/special-props.html they recommend using a prop called `id`

TL;DR `key` is essentially a reserved keyword that is not treated the same as other props.
